### PR TITLE
[FLINK-17691][Connectors/Kafka] Fix transactionId kafka string length bug

### PR DESCRIPTION
**Pull request by John-J-Mathews**
## What is the purpose of the change

The purpose of this change is to prevent Flink from failing to produce records in the case where the taskName exceeds 32,000 characters in length which is larger than the max transactionId kafka supports. 

## Brief change log

Truncate taskName to length 10_000 when constructing the transactionalIdGenerator.

## Verifying this change

create a sink kafka job in Semantic.EXACTLY_ONCE mode
make the operatorName's length large than 32767 (or a large sql operator, which sql string's length is larger than 32767)
```
FlinkKafkaProducer011 init TransactionalIdsGenerator prefix
 getRuntimeContext().getTaskName() + "-" + ((StreamingRuntimeContext) getRuntimeContext()).getOperatorUniqueID()
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) (No) 
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) (No)
  - The serializers: (yes / no / don't know) (No) 
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) (No) 
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know) (No) 
  - The S3 file system connector: (yes / no / don't know) (No) 

## Documentation

  - Does this pull request introduce a new feature? (yes / no) (No)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java [hunk 1]
```diff
@@ -145,6 +145,12 @@
 
 	private static final long serialVersionUID = 1L;
 
+	/**
+	 * Number of characters to truncate the taskName to for the Kafka transactionalId.
+	 * The maximum this can possibly be set to is 32,767 - (length of operatorUniqueId).
+	 */
+	private static final short maxTaskNameSize = 10_000;
```
Comment (John-J-Mathews) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:152]: Wasn't sure what to go with as a value here but using the max characters seems unnecessarily expensive and probably doesn't provide a ton of signal. 

If others agree, I thinking moving to something even lower may even be better than 10k given I believe the transactionId is sent over the network to the broker which is a lot of wasted network overhead.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java [hunk 2]
```diff
@@ -1084,8 +1090,14 @@ public void initializeState(FunctionInitializationContext context) throws Except
 			migrateNextTransactionalIdHindState(context);
 		}
 
+		String taskName = getRuntimeContext().getTaskName();
+		// Kafka transactional IDs are limited in length to be less than the max value of a short,
+		// so we truncate here if necessary to a more reasonable length string.
+		if (taskName.length() > maxTaskNameSize) {
+			taskName = taskName.substring(0, maxTaskNameSize);
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:1097]:
```suggestion
			taskName = taskName.substring(0, maxTaskNameSize);
			LOG.warn("Truncated task name for Kafka TransactionalId from {} to {}.", getRuntimeContext().getTaskName(), taskName);
```

I hope this compiles, just typed it out here in Github...

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java (hunk 1 above)
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:152]: Yes, I think even a string length of 10k is excessive. Let's maybe limit it to 1k?

**Review by aljoscha — CHANGES_REQUESTED** [review]
Thanks for the quick PR. I left some inline suggestions but I think it's mergeable after that.

**Review by aljoscha — APPROVED** [review]
The changes look good! I think we're also safe when it comes to savepoint compatibility because we were storing already-used IDs in state, so we can retrieve them again.

**Comment by aljoscha** [comment]
Thanks for the fix! I merged this.

**Pull request closed without merging**